### ISSUE TITLE
Editor leaks memory (again)

### ISSUE THREAD

**Issue by phlax** [translate/pootle#4916]
Profiling the editor it seems there are again issues with mem leaks. In fact even without clicking (just wiggling the mouse) its possible to farm memory.

I havent as yet got to the root of the problem

**Comment by phlax** [comment]
I believe the problem is to do with leaking nodes, perhaps in a react component.

I undertook the following test

- switched off tipsy and odometer
- reloaded
- garbage collected
- started profiling
- click back and forth between the same 2 units
- garbage collected
- click back and forth between the same 2 units
- garbage collected

This is a screenshot of the results

![mem-profile-same-two](https://cloud.githubusercontent.com/assets/454682/16915979/6a41bad6-4cf2-11e6-8ab1-f8f7e574aa34.png)

basically the nodecount increases by around 10 every time a unit is focused, and this seems to be what is leaking the memory

**Comment by julen** [comment]
Depending on the version of autosize in use, jackmoore/autosize#270 can be related (the fix was released in version 3.0.14).

**Comment by phlax** [comment]
apologies - i disabled autosize too in the above test.

I have also now disabled updateRelativeDates, so not much left running, but i still get a small increase in uncollected nodes on unit focus - and a ~significant increase in mem

**Comment by iafan** [comment]
Just thinking out loud: if there was no such a memory use pattern before, probably `git bisect` could help? Also, it's worth checking this pattern against different browsers — this might be related to how GC works on a particular browser (and changes between browser versions) rather than to the changes in JavaScript code itself. I tried to profile JS heap on Chrome, did some navigation between units for ~50 times and didn't see any indication that the memory is leaking.

**Comment by phlax** [comment]
I believe this if fixed - although the performance of the editor is still a bit of an issue - ive opened a related ticket https://github.com/translate/pootle/issues/4941

**Comment by julen** [comment]
I re-ran a profile on current master (switching between 2 consecutive units and forcing GC) and this is what I get:
![screen shot 2016-07-22 at 10 49 40](https://cloud.githubusercontent.com/assets/16768/17051738/6dd5bffc-4ffa-11e6-83a4-ecc165e8e717.png)

Is it in line with other people's?

**Comment by phlax** [comment]
if i gc - i return almost to baseline (in master) - with pr - i return to baseline

**Comment by julen** [comment]
Okay here are a couple of updated screenshots because I was using a debug build and the screenshot was missing the important part that was auto-cropped by Chrome.

The same test in current master:
![screen shot 2016-07-22 at 11 41 03](https://cloud.githubusercontent.com/assets/16768/17053212/3ad8428a-5001-11e6-98a7-92ca5d6df71d.png)

And this is with master prior to this change:
![screen shot 2016-07-22 at 11 37 05](https://cloud.githubusercontent.com/assets/16768/17053120/c58438cc-5000-11e6-989c-d3e727b39c0a.png)

So yeah the node count gets back to the initial value now as opposed to before, which was slightly higher. This of course has a bigger effect over time.

Well done on the tireless effort @phlax! 👍
